components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,384 @@
+                Err(Error::PackageNotFound("".to_string()))
+            }
+            Err(e) => {
+                debug!("error fetching ident {}: {:?}", ident, e);
+                Err(e)
+            }
+        }
+    }
+
+    // This function and it's sibling get_cached_artifact in
+    // install.rs deserve to be refactored to eke out commonality.
+    /// This ensures the identified package is in the local cache,
+    /// verifies it, and returns a handle to the package's metadata.
+    fn get_cached_archive<T>(&self,
+                             ui: &mut T,
+                             package: &PackageIdentTarget)
+                             -> Result<PackageArchive>
+        where T: UIWriter
+    {
+        let fetch_artifact = || self.fetch_artifact(ui, package);
+        if self.is_artifact_cached(package) {
+            debug!("Found {} in artifact cache, skipping remote download",
+                   package.ident);
+        } else if let Err(err) = retry(delay::Fixed::from(RETRY_WAIT).take(RETRIES), fetch_artifact)
+        {
+            return Err(Error::DownloadFailed(format!("We tried {} times but \
+                                                      could not download {}. \
+                                                      Last error was: {}",
+                                                     RETRIES, package, err)));
+        }
+
+        // At this point the artifact is in the cache...
+        let mut artifact = PackageArchive::new(self.cached_artifact_path(package));
+        ui.status(Status::Verifying, artifact.ident()?)?;
+        self.verify_artifact(ui, package, &mut artifact)?;
+        Ok(artifact)
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    /// Retrieve the identified package from the depot, ensuring that
+    /// the artifact is cached locally.
+    fn fetch_artifact<T>(&self, ui: &mut T, package: &PackageIdentTarget) -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading, package)?;
+        match self.api_client
+                  .fetch_package((&package.ident, package.target),
+                                 self.token,
+                                 self.artifact_cache_path,
+                                 ui.progress())
+        {
+            Ok(_) => Ok(()),
+            Err(api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
+                println!("Host platform or architecture not supported by the targeted depot; \
+                          skipping.");
+                Ok(())
+            }
+            Err(e) => Err(Error::from(e)),
+        }
+    }
+
+    fn fetch_origin_key<T>(&self,
+                           ui: &mut T,
+                           name_with_rev: &str,
+                           token: Option<&str>)
+                           -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading,
```

Comment:
It would be nice to also have a get_cached_origin_key function (analogous to the get_cached_artifact fn)